### PR TITLE
Refactor Cache Code

### PR DESCRIPTION
**Pull request by lunaruan**
This PR refactors the cache code by moving it out of `ReactFiberCacheComponent` to `ReactFiberTransitionPool` in anticipation of it being reused by multiple stacks (ie. transition tracing)

### CODE REVIEW

**Review by acdlite — COMMENTED** [review]
Overall this PR makes sense!

One nit regarding naming: The word "pool" refers to how Cache implementation reuses caches across multiple pending transitions. Kind of like a thread pool. It doesn't really apply in the case of transition tracing. So I would combine the new ReactFiberTransitionPool module with the existing ReactFiberTransition module, and rename `pushTransitionPool` to just `pushTransition`, etc.

File: packages/react-reconciler/src/ReactFiberTransition.new.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  // Otherwise, check the root's cache pool.
+  const root = (getWorkInProgressRoot(): any);
+  const cacheFromRootCachePool = root.pooledCache;
+
+  return cacheFromRootCachePool;
+}
+
+export function requestCache(renderLanes: Lanes): Cache {
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberTransition.new.js:50]: "pool" still makes sense here because that's where this function requests the cache from. So I would rename this one back to `requestCacheFromPool`, and `peekCache` back to `peekCacheFromPool`.

**Review by acdlite — APPROVED** [review]
Looks great, thanks!

File: packages/react-reconciler/src/ReactFiberTransitionPool.old.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
+/**
```
Comment (tjallingt) [inline packages/react-reconciler/src/ReactFiberTransitionPool.old.js:1]: Is this file intentional? I think it's unused since the contents have been moved to the ReactFiberTransition file.
